samza-kv-rocksdb/src/main/scala/org/apache/samza/storage/kv/RocksDbKeyValueStore.scala

```diff
@@ -164,55 +173,79 @@ class RocksDbKeyValueStore(
     metrics.deletes.inc(deletes)
   }
 
-  def delete(key: Array[Byte]) {
+  def delete(key: Array[Byte]): Unit = withStateCheck {
     metrics.deletes.inc
     put(key, null)
   }
 
-  def range(from: Array[Byte], to: Array[Byte]): KeyValueIterator[Array[Byte], Array[Byte]] = {
+  def range(from: Array[Byte], to: Array[Byte]): KeyValueIterator[Array[Byte], Array[Byte]] = withStateCheck {
     metrics.ranges.inc
     require(from != null && to != null, "Null bound not allowed.")
     new RocksDbRangeIterator(db.newIterator(), from, to)
   }
 
-  def all(): KeyValueIterator[Array[Byte], Array[Byte]] = {
+  def all(): KeyValueIterator[Array[Byte], Array[Byte]] = withStateCheck {
     metrics.alls.inc
     val iter = db.newIterator()
     iter.seekToFirst()
     new RocksDbIterator(iter)
   }
 
-  def flush {
+  def flush(): Unit = withStateCheck {
     metrics.flushes.inc
     trace("Flushing store: %s" format storeName)
     db.flush(flushOptions)
     trace("Flushed store: %s" format storeName)
   }
 
-  def close() {
-    trace("Closing.")
-    db.close()
+  def close(): Unit = {
+    stateChangeLock.writeLock().lock()
+    try {
+      trace("Closing.")
+      if (stackAtFirstClose == null) { // first close
+        stackAtFirstClose = new Exception()
+        db.close()
+      } else {
+        warn(new SamzaException("Close called again on a closed store: %s. Ignoring this close." +
+          "Stack at first close is under 'Caused By'." format storeName, stackAtFirstClose))
+      }
+    } finally {
+      stateChangeLock.writeLock().unlock()
+    }
+  }
+
+  private def withStateCheck[T](fn: => T): T = {
```

Comment:
Minor: cleaner to call it `withCloseGuard` perhaps?

Reply:
How about `ifOpen` or something?

Reply:
yeah, even better!